src/Disks/S3/DiskS3.cpp

```diff
@@ -996,15 +995,14 @@ void DiskS3::restoreFileOperations(const RestoreInformation & restore_informatio
 
             LOG_TRACE(log, "Move directory to 'detached' {} -> {}", path, detached_path);
 
-            fs::path from_path = fs::path(metadata_path) / path;
-            fs::path to_path = fs::path(metadata_path) / detached_path;
+            fs::path from_path = fs::path(path);
+            fs::path to_path = fs::path(detached_path);
             if (path.ends_with('/'))
                 to_path /= from_path.parent_path().filename();
             else
                 to_path /= from_path.filename();
-            fs::create_directories(to_path);
-            fs::copy(from_path, to_path, fs::copy_options::recursive | fs::copy_options::overwrite_existing);
-            fs::remove_all(from_path);
+
+            metadata_disk->moveDirectory(from_path, to_path);
```

Comment:
It will be more correct to make it the same as it was because there is a difference. `moveDirectory` will fail after this change if `to_path` is non-empty. But previously it overwrited existing, which is good because will still work in case for example abrupt restart.

Also please fix build check:
```
2021-11-12 04:16:55 /build/obj-x86_64-linux-gnu/../src/Disks/DiskLocal.cpp:117:5: error: use '= default' to define a trivial default constructor [modernize-use-equals-default,-warnings-as-errors]
2021-11-12 04:16:55     DiskLocalDirectoryIterator() {}
2021-11-12 04:16:55     ^                            ~~
2021-11-12 04:16:55                                  = default;
2021-11-12 04:16:55 64744 warnings generated.
2021-11-12 04:16:55 Suppressed 64743 warnings (64742 in non-user code, 1 with check filters).
```
https://clickhouse-builds.s3.yandex.net/31136/f7078a17ce8b63c0a0e76887d175b58ba0a4614d/clickhouse_special_build_check/report.html#fail1

Reply:
I think copy files and directories from 'from_path' to 'to_path' is little slow than rename. As you mentioned rename requires the destination is an empty directory, so why not clear all in 'to_path' directory?

Reply:
Yes, lets do this way.